### ISSUE TITLE
Borrow error when yielding with nested generators 

### ISSUE THREAD

**Issue by da-x** [rust-lang/rust#45093]
Hi, I got some feedback regarding the generator feature that is currently in nightly.

I am testing whether it is possible to yield out of a generator while another generator is in scope, for example:

```rust
fn main() {
    let mut generator = || {
        yield 1;

        let mut sub_generator = || {
            yield 2;
        };

        match sub_generator.resume() {
            GeneratorState::Yielded(x) => {
                yield x;
            }
            _ => panic!("unexpected value from resume"),
        }

        yield 3;

        return "foo"
    };

    match generator.resume() {
        GeneratorState::Yielded(1) => {}
        _ => panic!("unexpected value from resume"),
    }
    match generator.resume() {
        GeneratorState::Yielded(2) => {}
        _ => panic!("unexpected value from resume"),
    }
    match generator.resume() {
        GeneratorState::Yielded(3) => {}
        _ => panic!("unexpected value from resume"),
    }
    match generator.resume() {
        GeneratorState::Complete("foo") => {}
        _ => panic!("unexpected value from resume"),
    }
}
```

However I am getting:

```
error[E0626]: borrow may still be in use when generator yields
  --> src/main.rs:17:15
   |
17 |         match sub_generator.resume() {
   |               ^^^^^^^^^^^^^
...
24 |         yield 3;
   |         ------- possible yield occurs here
```

rustc: `rustc 1.22.0-nightly (05cbece09 2017-10-06)`

I am not sure what borrow this error is talking about. 

Is `sub_generator` not owned?

Is there an intention to support such use cases?

**Comment by sinkuu** [comment]
This could be a bug that rustc is pointing out a wrong `yield`. If I comment out `yield x;`, which is obviously during `sub_generator` borrow, E0626 disappears.

**Comment by da-x** [comment]
#43122 @aturon

**Comment by Zoxc** [comment]
The `resume` function borrows the generator and that borrow is considered to be in live for the entire match when we are looking for yields which conflicts with it. A workaround is to move that call outside the match.
```rust
let result = sub_generator.resume();
match result {
    GeneratorState::Yielded(x) => {
        yield x;
    }
    _ => panic!("unexpected value from resume"),
}
```

**Comment by da-x** [comment]
I see. 

Is this somewhat related to https://github.com/rust-lang/rust/issues/44100 ? I guess that the borrow checker could have narrowed the borrowing to the matched value.

**Comment by Zoxc** [comment]
@da-x No, it's the fault of the code which finds out which values could be stored in the generator.

**Comment by Arnavion** [comment]
Duplicate of #44197

**Comment by Zoxc** [comment]
Fixed by #45337.
